internal/terraform/evaluate.go

```diff
@@ -793,6 +798,127 @@ func (d *evaluationStateData) GetResource(addr addrs.Resource, rng tfdiags.Sourc
 	return ret, diags
 }
 
+func (d *evaluationStateData) getListResource(config *configs.Resource, rng tfdiags.SourceRange) (cty.Value, tfdiags.Diagnostics) {
+	var diags tfdiags.Diagnostics
+	switch d.Operation {
+	case walkValidate:
+		return cty.DynamicVal, diags
+	case walkQuery:
+		// continue
+	default:
+		return cty.DynamicVal, diags.Append(&hcl.Diagnostic{
+			Severity: hcl.DiagError,
+			Summary:  `Unsupported operation`,
+			Detail:   fmt.Sprintf("List resources are not supported in %s operations.", d.Operation),
+			Subject:  rng.ToHCL().Ptr(),
+		})
+	}
+
+	// The provider result of a list resource is always a tuple, but
+	// we will wrap that tuple in an object with a single attribute "data",
+	// so that we can differentiate between a list resource instance (list.aws_instance.test[index])
+	// and the elements of the result of a list resource instance (list.aws_instance.test.data[index])
+	wrappedVal := func(v *states.ResourceInstanceObject) cty.Value {
+		return cty.ObjectVal(map[string]cty.Value{"data": v.Value})
+	}
+	lAddr := config.Addr()
+	mAddr := addrs.Resource{
+		Mode: addrs.ManagedResourceMode,
+		Type: lAddr.Type,
+		Name: lAddr.Name,
+	}
+	schema := d.getResourceSchema(mAddr, config.Provider)
```

Comment:
Could we rename this variable to clarify the type of schema (and the implied type) we're dealing with? At first, I was unsure whether it was a resource or a list schema.